File: examples/react-native/example/ios/SessionReplayPluginModule.swift
//
//  SessionReplayPluginModule.swift
//  example
//
//  Created by Curtis Liu on 10/7/24.
//
import Foundation
import AmplitudeSessionReplay


@objc(SessionReplayPluginModule)
class SessionReplayPluginModule: NSObject {
  
  var sessionReplay: SessionReplay!
  
  @objc(setup:deviceId:sessionId:resolve:reject:)
  func setup(_ apiKey: String, deviceId: String, sessionId: NSNumber, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
    print("setup: \(apiKey) \(deviceId) \(sessionId)")
    sessionReplay = SessionReplay(apiKey:apiKey,
                                  deviceId: deviceId,
                                  sessionId: sessionId.int64Value,
                                  sampleRate: 1.0,
                                  logger:ConsoleLogger(logLevel: LogLevelEnum.DEBUG.rawValue),
                                  enableRemoteConfig: false)
    sessionReplay.start()
    resolve(nil)
  }
  
  @objc(setSessionId:resolve:reject:)
  func setSessionId(_ sessionId: NSNumber, resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
    print("setSessionId: \(sessionId)")
    sessionReplay.sessionId = sessionId.int64Value
    resolve(nil)
  }
  
  @objc(getSessionId:reject:)
  func getSessionId(_ resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
    print("getSessionId")
    resolve(NSNumber(value:sessionReplay.sessionId))
  }
  
  @objc(getSessionReplayProperties:reject:)
  func getSessionReplayProperties(_ resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) {
    print("getSessionReplayProperties")
    resolve(sessionReplay.additionalEventProperties)
  }
  
  @objc(flush:reject:)
  func flush(_ resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
    print("flush")
    sessionReplay.flush()
    resolve(nil)
  }
  
  @objc(teardown:reject:)
  func teardown(_ resolve: RCTPromiseResolveBlock, reject: RCTPromiseRejectBlock) -> Void {
    print("teardown")
    sessionReplay.stop()
    resolve(nil)
  }
  
}

File: examples/react-native/example/android/app/src/main/java/com/example/SessionReplayPluginModule.kt
package com.example

import com.amplitude.android.sessionreplay.SessionReplay
import com.amplitude.common.Logger
import com.amplitude.common.android.LogcatLogger
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.WritableMap
import com.facebook.react.bridge.WritableNativeMap

class SessionReplayPluginModule(private val reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {
    private lateinit var sessionReplay: SessionReplay

    override fun getName(): String {
        return "SessionReplayPluginModule"
    }

    @ReactMethod
    fun setup(apiKey: String, deviceId: String?, sessionId: Double) {
        LogcatLogger.logger.logMode = Logger.LogMode.DEBUG
        sessionReplay = SessionReplay(
            apiKey,
            reactContext.applicationContext,
            deviceId ?: "",
            sessionId.toLong(),
            logger = LogcatLogger.logger,
            sampleRate = 1.0,
            enableRemoteConfig = false,
        )
    }

    @ReactMethod
    fun setSessionId(sessionId: Double) {
        sessionReplay.setSessionId(sessionId.toLong())
    }

    @ReactMethod
    fun getSessionId(promise: Promise) {
        promise.resolve(sessionReplay.getSessionId().toDouble())
    }

    @ReactMethod
    fun getSessionReplayProperties(promise: Promise) {
        val properties: Map<String, Any> = sessionReplay.getSessionReplayProperties()
        val map: WritableMap = WritableNativeMap()
        for ((key, value) in properties) {
            if (value is String) {
                map.putString(key, value)
            } else if (value is Int) {
                map.putInt(key, value)
            } else if (value is Long) {
                map.putDouble(key, value.toDouble())
            } else if (value is Double) {
                map.putDouble(key, value)
            } else if (value is Boolean) {
                map.putBoolean(key, value)
            }
        }
        promise.resolve(map)
    }

    @ReactMethod
    fun flush() {
        sessionReplay.flush()
    }

    @ReactMethod
    fun teardown() {
        sessionReplay.shutdown()
    }

    override fun invalidate() {
        sessionReplay.shutdown()
    }
}
